File: k3builder/internal/controller/exposedapp_controller.go
/*
Copyright 2025.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package controller

import (
	"context"
	"fmt"
	"k8s.io/apimachinery/pkg/types"
	"k8s.io/client-go/tools/record"

	"k8s.io/apimachinery/pkg/api/meta"
	"k8s.io/apimachinery/pkg/runtime"
	ctrl "sigs.k8s.io/controller-runtime"
	"sigs.k8s.io/controller-runtime/pkg/client"
	"sigs.k8s.io/controller-runtime/pkg/log"

	stablev1 "k3builder.com/exposedapp/api/v1"
	appsv1 "k8s.io/api/apps/v1"
	corev1 "k8s.io/api/core/v1"
	apierrors "k8s.io/apimachinery/pkg/api/errors"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

// ExposedAppReconciler reconciles a ExposedApp object
type ExposedAppReconciler struct {
	client.Client
	Scheme   *runtime.Scheme
	Recorder record.EventRecorder
	PodName  string
}

const mainContainerName = "main"

// +kubebuilder:rbac:groups=stable.k3builder.com,resources=exposedapps,verbs=get;list;watch;create;update;patch;delete
// +kubebuilder:rbac:groups=stable.k3builder.com,resources=exposedapps/status,verbs=get;update;patch
// +kubebuilder:rbac:groups=stable.k3builder.com,resources=exposedapps/finalizers,verbs=update
// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=create;update;delete;get;watch;list
// +kubebuilder:rbac:groups=core,resources=services,verbs=create;update;delete;get;watch;list
// +kubebuilder:rbac:groups=core,resources=events,verbs=create;patch

func (r *ExposedAppReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
	logger := log.FromContext(ctx)

	exposedApp := &stablev1.ExposedApp{}
	namespace := req.NamespacedName.Namespace
	err := r.Get(ctx, req.NamespacedName, exposedApp)
	if err != nil {
		if apierrors.IsNotFound(err) {
			logger.Info("ExposedApp resource not found, probably deleted.")
			return ctrl.Result{}, nil
		}
		return ctrl.Result{}, err
	}

	if len(exposedApp.Status.Conditions) == 0 {
		err = r.InitStatus(ctx, exposedApp)
		if err != nil {
			return ctrl.Result{}, err
		}
	}

	deploymentName := fmt.Sprintf("%s-%s-deployment", exposedApp.Name, exposedApp.Namespace)
	serviceName := fmt.Sprintf("%s-%s-service", exposedApp.Name, exposedApp.Namespace)

	deployment := &appsv1.Deployment{}
	err = r.Get(ctx, types.NamespacedName{Name: deploymentName, Namespace: namespace}, deployment)
	if err != nil {
		if apierrors.IsNotFound(err) {
			deployment, err = r.Deployment(exposedApp, deploymentName)
			if err != nil {
				return ctrl.Result{}, err
			}
			err = r.Create(ctx, deployment)
			if err != nil {
				return ctrl.Result{}, err
			}
			r.Recorder.Eventf(exposedApp, "Normal",
				"DeploymentCreated",
				"Deployment %s created",
				deploymentName)
		} else {
			return ctrl.Result{}, err
		}
	} else {
		deployment.Spec.Replicas = &exposedApp.Spec.Replicas
		deployment.Spec.Template.Spec.Containers[0].Image = exposedApp.Spec.Image
		deployment.Spec.Template.Spec.Containers[0].Ports[0].ContainerPort = exposedApp.Spec.ContainerPort
		deployment.Spec.Template.Spec.Containers[0].Ports[0].Protocol = corev1.Protocol(exposedApp.Spec.Protocol)
		err = r.Update(ctx, deployment)
		if err != nil {
			return ctrl.Result{}, err
		}
		r.Recorder.Eventf(exposedApp, "Normal",
			"DeploymentUpdated",
			"Deployment %s updated",
			deploymentName)
	}
	meta.SetStatusCondition(&exposedApp.Status.Conditions, metav1.Condition{
		Status:  metav1.ConditionTrue,
		Reason:  "Provisioned",
		Message: "Deployment ready",
		Type:    "DeploymentReady",
	})
	err = r.Status().Update(ctx, exposedApp)
	if err != nil {
		logger.Error(err, "Unable to update ExposedApp Status")
		return ctrl.Result{}, err
	}

	service := &corev1.Service{}
	err = r.Get(ctx, types.NamespacedName{Name: serviceName, Namespace: namespace}, service)
	if err != nil {
		if apierrors.IsNotFound(err) {
			service, err = r.Service(exposedApp, deployment.Spec.Template.ObjectMeta.Labels, serviceName)
			if err != nil {
				return ctrl.Result{}, err
			}
			err = r.Create(ctx, service)
			if err != nil {
				return ctrl.Result{}, err
			}
			r.Recorder.Eventf(exposedApp, "Normal",
				"ServiceCreated",
				"Service %s created",
				serviceName)
		} else {
			return ctrl.Result{}, err
		}
	} else {
		service.Spec.Ports[0].Port = exposedApp.Spec.Port
		service.Spec.Ports[0].Protocol = corev1.Protocol(exposedApp.Spec.Protocol)
		if exposedApp.Spec.NodePort != nil {
			service.Spec.Ports[0].NodePort = *exposedApp.Spec.NodePort
		}
		if len(exposedApp.Spec.ServiceType) != 0 {
			service.Spec.Type = corev1.ServiceType(exposedApp.Spec.ServiceType)
		}
		err = r.Update(ctx, service)
		if err != nil {
			return ctrl.Result{}, err
		}
		r.Recorder.Eventf(exposedApp, "Normal",
			"ServiceUpdated",
			"Service %s updated",
			serviceName)
	}

	exposedApp.Status.ServiceName = serviceName
	exposedApp.Status.DeploymentName = deploymentName
	exposedApp.Status.LastUpdateBy = r.PodName
	exposedApp.Status.Ready = fmt.Sprintf("%d/%d", deployment.Status.ReadyReplicas, deployment.Status.ReadyReplicas)
	meta.SetStatusCondition(&exposedApp.Status.Conditions, metav1.Condition{
		Status:  metav1.ConditionTrue,
		Reason:  "Provisioned",
		Message: "Service ready",
		Type:    "ServiceReady",
	})
	err = r.Status().Update(ctx, exposedApp)
	if err != nil {
		logger.Error(err, "Unable to update ExposedApp Status")
		return ctrl.Result{}, err
	}

	// RequeueAfter not required here as Deployment is a dependent resource and this controller watches for changes
	// RequeueAfter might be useful for managing external resources, like AWS ALB to poll the status
	return ctrl.Result{}, nil
}

func (r *ExposedAppReconciler) InitStatus(ctx context.Context, exposedApp *stablev1.ExposedApp) error {
	logger := log.FromContext(ctx)
	meta.SetStatusCondition(&exposedApp.Status.Conditions, metav1.Condition{
		Status:  metav1.ConditionFalse,
		Reason:  "Provisioning",
		Message: "Deployment not ready",
		Type:    "DeploymentReady",
	})
	meta.SetStatusCondition(&exposedApp.Status.Conditions, metav1.Condition{
		Status:  metav1.ConditionFalse,
		Reason:  "Provisioning",
		Message: "Service not ready",
		Type:    "ServiceReady",
	})
	err := r.Status().Update(ctx, exposedApp)
	if err != nil {
		logger.Error(err, "Unable to update ExposedApp Status")
		return err
	}
	return nil
}

func (r *ExposedAppReconciler) Service(exposedApp *stablev1.ExposedApp, podLabels map[string]string, serviceName string) (*corev1.Service, error) {
	servicePort := corev1.ServicePort{
		Protocol: corev1.Protocol(exposedApp.Spec.Protocol),
		Port:     exposedApp.Spec.Port,
	}
	if exposedApp.Spec.NodePort != nil {
		servicePort.NodePort = *exposedApp.Spec.NodePort
	}
	service := &corev1.Service{
		ObjectMeta: metav1.ObjectMeta{
			Namespace: exposedApp.Namespace,
			Name:      serviceName,
		},
		Spec: corev1.ServiceSpec{
			Selector: podLabels,
			Ports:    []corev1.ServicePort{servicePort},
		},
	}
	if len(exposedApp.Spec.ServiceType) != 0 {
		service.Spec.Type = corev1.ServiceType(exposedApp.Spec.ServiceType)
	}
	// Set ExposedApp as owner
	err := ctrl.SetControllerReference(exposedApp, service, r.Scheme)
	if err != nil {
		return nil, err
	}
	return service, nil
}

func (r *ExposedAppReconciler) Deployment(exposedApp *stablev1.ExposedApp, deploymentName string) (*appsv1.Deployment, error) {
	labels := map[string]string{"app.kubernetes.io/name": fmt.Sprintf("%s-%s", exposedApp.Name, exposedApp.Namespace)}
	deployment := &appsv1.Deployment{
		ObjectMeta: metav1.ObjectMeta{
			Namespace: exposedApp.Namespace,
			Name:      deploymentName,
		},
		Spec: appsv1.DeploymentSpec{
			Replicas: &exposedApp.Spec.Replicas,
			Selector: &metav1.LabelSelector{
				MatchLabels: labels,
			},
			Template: corev1.PodTemplateSpec{
				ObjectMeta: metav1.ObjectMeta{
					Labels: labels,
				},
				Spec: corev1.PodSpec{
					Containers: []corev1.Container{{
						Image: exposedApp.Spec.Image,
						Name:  mainContainerName,
						Ports: []corev1.ContainerPort{{
							ContainerPort: exposedApp.Spec.ContainerPort,
							Protocol:      corev1.Protocol(exposedApp.Spec.Protocol),
						}},
					}},
				},
			},
		},
	}
	// Set ExposedApp as owner
	err := ctrl.SetControllerReference(exposedApp, deployment, r.Scheme)
	if err != nil {
		return nil, err
	}
	return deployment, nil
}

// SetupWithManager sets up the controller with the Manager.
func (r *ExposedAppReconciler) SetupWithManager(mgr ctrl.Manager) error {
	return ctrl.NewControllerManagedBy(mgr).
		For(&stablev1.ExposedApp{}).
		Named("exposedapp").
		Owns(&appsv1.Deployment{}).
		Owns(&corev1.Service{}).
		Complete(r)
}

File: kube-rs-o6r/src/exposed_app_controller.rs
use crate::exposed_app::{ExposedApp, ExposedAppStatus};
use futures::StreamExt;
use k8s_openapi::api::apps::v1::{Deployment, DeploymentSpec};
use k8s_openapi::api::core::v1::{
    Container, ContainerPort, PodSpec, PodTemplateSpec, Service, ServicePort, ServiceSpec,
};
use k8s_openapi::apimachinery::pkg::apis::meta::v1::{Condition, LabelSelector, ObjectMeta, Time};
use k8s_openapi::apimachinery::pkg::util::intstr::IntOrString::Int;
use k8s_openapi::chrono::{DateTime, Utc};
use kube::api::Patch::Merge;
use kube::api::{PatchParams, PostParams};
use kube::runtime::Controller;
use kube::runtime::controller::Action;
use kube::runtime::events::{Event, EventType, Recorder, Reporter};
use kube::runtime::watcher::Config;
use kube::{Api, Client, Error, Resource};
use serde::Serialize;
use serde::de::DeserializeOwned;
use serde_json::json;
use std::collections::BTreeMap;
use std::fmt::Debug;
use std::sync::Arc;
use std::time::Duration;
use tracing::{info, warn};

#[derive(Clone)]
struct Data {
    client: Client,
    recorder: Recorder,
}

impl Data {
    pub fn new(client: Client, recorder: Recorder) -> Self {
        Self { client, recorder }
    }
}

async fn save<T: Clone + DeserializeOwned + Debug + Serialize>(
    api: &Api<T>,
    name: &str,
    object: &T,
) -> Result<T, Error> {
    match api.get_opt(name).await? {
        None => api.create(&PostParams::default(), &object).await,
        Some(_) => {
            api.patch(name, &PatchParams::default(), &Merge(&object))
                .await
        }
    }
}

async fn publish_event(
    recorder: &Recorder,
    exposed_app: &ExposedApp,
    event: Event,
) -> Result<(), Error> {
    let reference = exposed_app.object_ref(&());
    recorder.publish(&event, &reference).await
}

async fn reconcile(object: Arc<ExposedApp>, ctx: Arc<Data>) -> Result<Action, Error> {
    let namespace = object.metadata.namespace.clone().unwrap();
    let name = object.metadata.name.clone().unwrap();
    let generation = object.metadata.generation.unwrap();
    let mut conditions = object
        .status
        .clone()
        .unwrap_or(ExposedAppStatus::default())
        .conditions
        .unwrap_or(Vec::new());
    info!("Reconciling {} {}", name, namespace,);
    let services: Api<Service> = Api::namespaced(ctx.client.clone(), namespace.as_str());
    let deployments: Api<Deployment> = Api::namespaced(ctx.client.clone(), namespace.as_str());
    let exposed_apps: Api<ExposedApp> = Api::namespaced(ctx.client.clone(), namespace.as_str());
    let recorder = ctx.recorder.clone();
    let mut pod_labels: BTreeMap<String, String> = BTreeMap::new();
    pod_labels.insert(String::from("app.kubernetes.io/name"), name.clone());
    let deployment_name = format!("{}-deployment", name);
    let service_name = format!("{}-service", name);
    let new_deployment = deployment(
        deployment_name.as_str(),
        namespace.as_str(),
        &pod_labels,
        &object,
    );
    save(&deployments, deployment_name.as_str(), &new_deployment).await?;
    publish_event(
        &recorder,
        &object,
        Event {
            type_: EventType::Normal,
            reason: String::from("Deployment provisioning requested"),
            action: String::from("DeploymentProvisioned"),
            note: Some(format!("Deployment {} provisioned", deployment_name)),
            secondary: None,
        },
    )
    .await?;
    patch_status(
        name.as_str(),
        &exposed_apps,
        ExposedAppStatus {
            deployment_name: Some(deployment_name.clone()),
            ..Default::default()
        },
    )
    .await?;
    conditions = set_condition(
        &conditions,
        Condition {
            last_transition_time: Time(DateTime::from(Utc::now())),
            message: "Deployment ready".into(),
            observed_generation: Some(generation),
            reason: "Provisioned".into(),
            status: "True".into(),
            type_: "DeploymentReady".into(),
        },
    );
    patch_conditions(name.as_str(), &exposed_apps, &conditions).await?;
    let new_service = service(
        service_name.as_str(),
        namespace.as_str(),
        &pod_labels,
        &object,
    );
    save(&services, service_name.as_str(), &new_service).await?;
    publish_event(
        &recorder,
        &object,
        Event {
            type_: EventType::Normal,
            reason: String::from("Service provisioning requested"),
            action: String::from("ServiceProvisioned"),
            note: Some(format!("Service {} provisioned", service_name)),
            secondary: None,
        },
    )
    .await?;
    patch_status(
        name.as_str(),
        &exposed_apps,
        ExposedAppStatus {
            service_name: Some(service_name.clone()),
            ..Default::default()
        },
    )
    .await?;
    conditions = set_condition(
        &conditions,
        Condition {
            last_transition_time: Time(DateTime::from(Utc::now())),
            message: "Service ready".into(),
            observed_generation: Some(generation),
            reason: "Provisioned".into(),
            status: "True".into(),
            type_: "ServiceReady".into(),
        },
    );
    patch_conditions(name.as_str(), &exposed_apps, &conditions).await?;
    Ok(Action::await_change())
}

async fn patch_status(
    name: &str,
    api: &Api<ExposedApp>,
    app_status: ExposedAppStatus,
) -> Result<ExposedApp, Error> {
    let status = json!({
        "status": app_status,
    });
    api.patch_status(name, &PatchParams::default(), &Merge(&status))
        .await
}

async fn patch_conditions(
    name: &str,
    api: &Api<ExposedApp>,
    conditions: &Vec<Condition>,
) -> Result<ExposedApp, Error> {
    let status = json!({
        "status": ExposedAppStatus {
            conditions: Some(conditions.clone()),
            ..Default::default()
        },
    });
    api.patch_status(name, &PatchParams::default(), &Merge(&status))
        .await
}

fn set_condition(conditions: &Vec<Condition>, condition: Condition) -> Vec<Condition> {
    let mut new_conditions = conditions.clone();
    for c in new_conditions.iter_mut() {
        if c.type_ == condition.type_ {
            *c = condition;
            return new_conditions;
        }
    }
    new_conditions.push(condition);
    new_conditions
}

fn deployment(
    name: &str,
    namespace: &str,
    pod_labels: &BTreeMap<String, String>,
    exposed_app: &ExposedApp,
) -> Deployment {
    let obj_ref = exposed_app.controller_owner_ref(&()).unwrap();
    Deployment {
        metadata: ObjectMeta {
            name: Some(String::from(name)),
            namespace: Some(String::from(namespace)),
            owner_references: Some(vec![obj_ref]),
            ..ObjectMeta::default()
        },
        spec: Some(DeploymentSpec {
            replicas: Some(exposed_app.spec.replicas),
            selector: LabelSelector {
                match_labels: Some(pod_labels.clone()),
                ..LabelSelector::default()
            },
            template: PodTemplateSpec {
                metadata: Some(ObjectMeta {
                    labels: Some(pod_labels.clone()),
                    ..ObjectMeta::default()
                }),
                spec: Some(PodSpec {
                    containers: vec![Container {
                        name: String::from("main"),
                        image: Some(exposed_app.spec.image.clone()),
                        ports: Some(vec![ContainerPort {
                            protocol: Some(exposed_app.spec.protocol.clone()),
                            container_port: exposed_app.spec.container_port,
                            ..ContainerPort::default()
                        }]),
                        ..Container::default()
                    }],
                    ..PodSpec::default()
                }),
                ..PodTemplateSpec::default()
            },
            ..DeploymentSpec::default()
        }),
        ..Deployment::default()
    }
}

fn service(
    name: &str,
    namespace: &str,
    pod_labels: &BTreeMap<String, String>,
    exposed_app: &ExposedApp,
) -> Service {
    let obj_ref = exposed_app.controller_owner_ref(&()).unwrap();
    Service {
        metadata: ObjectMeta {
            name: Some(String::from(name)),
            namespace: Some(String::from(namespace)),
            owner_references: Some(vec![obj_ref]),
            ..ObjectMeta::default()
        },
        spec: Some(ServiceSpec {
            selector: Some(pod_labels.clone()),
            type_: exposed_app.spec.service_type.clone(),
            ports: Some(vec![ServicePort {
                protocol: Some(exposed_app.spec.protocol.clone()),
                node_port: exposed_app.spec.node_port,
                target_port: Some(Int(exposed_app.spec.container_port)),
                port: exposed_app.spec.port,
                ..ServicePort::default()
            }]),
            ..ServiceSpec::default()
        }),
        ..Service::default()
    }
}

fn error_policy(_object: Arc<ExposedApp>, _err: &Error, _ctx: Arc<Data>) -> Action {
    Action::requeue(Duration::from_secs(5))
}

pub async fn run_controller(client: Client) -> () {
    let reporter = Reporter {
        controller: String::from("exposed-app-controller"),
        instance: None,
    };
    let recorder = Recorder::new(client.clone(), reporter);
    let data = Arc::new(Data::new(client.clone(), recorder));
    let exposed_apps: Api<ExposedApp> = Api::all(client.clone());
    let deployments: Api<Deployment> = Api::all(client.clone());
    let services: Api<Service> = Api::all(client.clone());
    Controller::new(exposed_apps, Config::default())
        .owns(deployments, Config::default())
        .owns(services, Config::default())
        .run(reconcile, error_policy, Arc::clone(&data))
        .for_each(|res| async {
            match res {
                Ok(o) => info!("reconciled {:?}", o),
                Err(e) => warn!("reconcile failed: {:?}", e),
            }
        })
        .await;
}
